File: tools/test_kimi_key.mjs
#!/usr/bin/env node
/**
 * 测试 LLM API Key 是否有效。
 * 同时支持 OpenAI-compatible（Moonshot）和 Anthropic-compatible（Kimi Code）协议。
 *
 * 用法：
 *   node tools/test_kimi_key.mjs
 *
 * 或指定其他 env 文件：
 *   PLAYTEST_ENV=/path/to/.playtest.env node tools/test_kimi_key.mjs
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

function loadPlaytestEnv() {
  let envPath = process.env.PLAYTEST_ENV;
  if (!envPath) {
    const repoRoot = path.resolve(__dirname, '..');
    envPath = path.join(repoRoot, 'packages', 'backend', '.playtest.env');
  }

  if (!fs.existsSync(envPath)) {
    console.error(`❌ 找不到配置文件: ${envPath}`);
    console.error('请先复制 packages/backend/.playtest.env.example 为 .playtest.env 并填入 API Key');
    process.exit(1);
  }

  const env = {};
  const text = fs.readFileSync(envPath, 'utf-8');
  for (const line of text.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const idx = trimmed.indexOf('=');
    if (idx === -1) continue;
    env[trimmed.slice(0, idx).trim()] = trimmed.slice(idx + 1).trim();
  }
  return env;
}

function detectProtocol(baseUrl) {
  const lower = baseUrl.toLowerCase();
  if (lower.includes('kimi.com') || lower.includes('anthropic')) return 'anthropic';
  return 'openai';
}

async function testOpenai(baseUrl, apiKey, model) {
  console.log('协议: OpenAI-compatible');

  // 测试 1: 获取模型列表
  console.log('测试 1/2: 获取模型列表...');
  const modelsRes = await fetch(`${baseUrl}/models`, {
    headers: { Authorization: `Bearer ${apiKey}` },
  });
  const modelsData = await modelsRes.json();
  if (!modelsRes.ok) {
    throw new Error(`/models 失败: ${modelsRes.status} ${JSON.stringify(modelsData)}`);
  }
  const modelIds = modelsData.data?.map((m) => m.id) || [];
  console.log(`✅ /models 成功，可用模型数: ${modelIds.length}`);
  if (modelIds.includes(model)) {
    console.log(`✅ 目标模型 '${model}' 在可用列表中`);
  } else {
    console.log(`⚠️ 目标模型 '${model}' 不在可用列表中，可用模型: ${modelIds.slice(0, 5).join(', ')}...`);
  }

  // 测试 2: 调用 chat.completions
  console.log('\n测试 2/2: 调用 chat.completions...');
  const chatRes = await fetch(`${baseUrl}/chat/completions`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      messages: [
        { role: 'system', content: '你是大富翁4玩家' },
        { role: 'user', content: '1+1等于多少？只回答数字' },
      ],
      max_completion_tokens: 50,
    }),
  });
  const chatData = await chatRes.json();
  if (!chatRes.ok) {
    throw new Error(`chat.completions 失败: ${chatRes.status} ${JSON.stringify(chatData)}`);
  }
  const content = chatData.choices?.[0]?.message?.content;
  console.log('✅ chat.completions 成功');
  console.log(`回复: ${content}`);
}

async function testAnthropic(baseUrl, apiKey, model) {
  console.log('协议: Anthropic-compatible（Kimi Code）');

  // Anthropic 协议没有 /models 列表可用，直接测 /v1/messages
  console.log('测试: 调用 /v1/messages...');
  const res = await fetch(`${baseUrl}/v1/messages`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-api-key': apiKey,
      'anthropic-version': '2023-06-01',
    },
    body: JSON.stringify({
      model,
      max_tokens: 50,
      system: '你是大富翁4玩家',
      messages: [{ role: 'user', content: '1+1等于多少？只回答数字' }],
    }),
  });
  const data = await res.json();
  if (!res.ok) {
    throw new Error(`/v1/messages 失败: ${res.status} ${JSON.stringify(data)}`);
  }
  const textBlock = data.content?.find((c) => c.type === 'text' || typeof c.text === 'string');
  const content = textBlock?.text;
  console.log('✅ /v1/messages 成功');
  console.log(`回复: ${content}`);
}

async function testKey() {
  const env = loadPlaytestEnv();
  const apiKey = env.PLAYTEST_LLM_API_KEY || '';
  const baseUrl = (env.PLAYTEST_LLM_BASE_URL || 'https://api.moonshot.cn/v1').replace(/\/$/, '');
  const model = env.PLAYTEST_LLM_MODEL || 'moonshot-v1-8k';

  if (!apiKey) {
    console.error('❌ PLAYTEST_LLM_API_KEY 为空');
    process.exit(1);
  }

  const protocol = detectProtocol(baseUrl);

  console.log(`base_url: ${baseUrl}`);
  console.log(`model: ${model}`);
  console.log(`key_prefix: ${apiKey.slice(0, 20)}...`);
  console.log(`key_length: ${apiKey.length}`);
  console.log('-'.repeat(40));

  try {
    if (protocol === 'anthropic') {
      await testAnthropic(baseUrl, apiKey, model);
    } else {
      await testOpenai(baseUrl, apiKey, model);
    }
  } catch (err) {
    console.error(`❌ ${err.message}`);
    process.exit(1);
  }

  console.log('\n✅ API Key 有效，可以开始 LLM playtest');
}

await testKey();

File: tools/test_kimi_key.py
#!/usr/bin/env python3
"""
测试 KIMI (Moonshot) API Key 是否有效。

用法：
    cd packages/backend
    cp .playtest.env.example .playtest.env
    # 编辑 .playtest.env 填入真实 key

    # 方式 1：直接运行（读取 packages/backend/.playtest.env）
    python3 ../../tools/test_kimi_key.py

    # 方式 2：指定 env 文件路径
    PLAYTEST_ENV=/path/to/.playtest.env python3 ../../tools/test_kimi_key.py

需要安装 openai 包：
    pip install openai
"""

import os
import sys
from pathlib import Path


def load_playtest_env():
    """读取 .playtest.env 文件。"""
    env_path = os.environ.get("PLAYTEST_ENV")
    if not env_path:
        # 默认读取 packages/backend/.playtest.env
        script_dir = Path(__file__).resolve().parent
        repo_root = script_dir.parent
        env_path = repo_root / "packages" / "backend" / ".playtest.env"

    env_path = Path(env_path)
    if not env_path.exists():
        print(f"❌ 找不到配置文件: {env_path}")
        print("请先复制 .playtest.env.example 为 .playtest.env 并填入 KIMI API Key")
        sys.exit(1)

    env = {}
    with open(env_path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if "=" not in line:
                continue
            key, value = line.split("=", 1)
            env[key.strip()] = value.strip()
    return env


def test_key():
    try:
        from openai import OpenAI
    except ImportError:
        print("❌ 未安装 openai 包，请先运行: pip install openai")
        sys.exit(1)

    env = load_playtest_env()
    api_key = env.get("PLAYTEST_LLM_API_KEY", "")
    base_url = env.get("PLAYTEST_LLM_BASE_URL", "https://api.moonshot.cn/v1")
    model = env.get("PLAYTEST_LLM_MODEL", "moonshot-v1-8k")

    if not api_key:
        print("❌ PLAYTEST_LLM_API_KEY 为空")
        sys.exit(1)

    print(f"base_url: {base_url}")
    print(f"model: {model}")
    print(f"key_prefix: {api_key[:20]}...")
    print(f"key_length: {len(api_key)}")
    print("-" * 40)

    client = OpenAI(api_key=api_key, base_url=base_url)

    try:
        # 先测试 /models 列表
        print("测试 1/2: 获取模型列表...")
        models = client.models.list()
        model_ids = [m.id for m in models.data]
        print(f"✅ /models 成功，可用模型数: {len(model_ids)}")
        if model in model_ids:
            print(f"✅ 目标模型 '{model}' 在可用列表中")
        else:
            print(f"⚠️ 目标模型 '{model}' 不在可用列表中，可用模型: {model_ids[:5]}...")
    except Exception as e:
        print(f"❌ /models 失败: {e}")
        sys.exit(1)

    try:
        # 再测试 chat.completions
        print("\n测试 2/2: 调用 chat.completions...")
        completion = client.chat.completions.create(
            model=model,
            messages=[
                {"role": "system", "content": "你是大富翁4玩家"},
                {"role": "user", "content": "1+1等于多少？只回答数字"},
            ],
            max_completion_tokens=50,
        )
        content = completion.choices[0].message.content
        print(f"✅ chat.completions 成功")
        print(f"回复: {content}")
    except Exception as e:
        print(f"❌ chat.completions 失败: {e}")
        sys.exit(1)

    print("\n✅ API Key 有效，可以开始 LLM playtest")


if __name__ == "__main__":
    test_key()
